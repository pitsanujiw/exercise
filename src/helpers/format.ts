import { cloneDeep, set, get, PropertyPath } from 'lodash'
import { formatCurrency } from '@coingecko/cryptoformat'
import numeral from 'numeral'

const numberFormatter = new Intl.NumberFormat('en-US')

const percentFormatter = new Intl.NumberFormat('en-US', {
  style: 'percent',
  minimumFractionDigits: 2,
  maximumFractionDigits: 2
})

const Format = {
  bigNumber: (value: number): string => numeral(value).format('($0.00a)'),

  currency: (value: number): string => formatCurrency(value, 'USD', 'en'),

  percent: (value: number): string => percentFormatter.format(value / 100),

  toNumbers: <T extends Object>(data: Array<T>, keys: Array<keyof T>): T[] => {
    const cloneData = cloneDeep(data)

    const newData: Array<T> = cloneData.map((object: any) => {
      keys.forEach(key => {
        const value = Number(object[key])
        set(object, key, value)
      })

      return object
    })

    return newData
  },

  toNumber: <T>(data: T, path: PropertyPath): number => Number(get(data, path))
}

export { Format, numberFormatter, percentFormatter }
