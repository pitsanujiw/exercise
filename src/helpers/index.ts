export * from './order-book-stream';
export * from './format';
export * from './test-utils';