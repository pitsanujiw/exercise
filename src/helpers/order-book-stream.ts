import { TOrder, TOrderBookFunction, TOrderBookStream, TSide } from "@types";
import faker from "faker";

class OrderBookStream {
  intervalId: number;
  subscribers: TOrderBookFunction[] = [];
  intervalRate = 3000;
  currentData: TOrderBookStream;
  userOrder: TOrderBookStream = {
    buy: [],
    sell: [],
  };

  constructor() {
    this.currentData = this.produceData();
    this.intervalId = window.setInterval(() => {
      this.subscribers.forEach((callback) => {
        this.currentData = this.produceData();
        callback(this.currentData);
      });
    }, this.intervalRate);
  }

  addOrder(side: TSide, order: TOrder) {
    this.userOrder[side].push(order);
  }

  produceData(): TOrderBookStream {
    const buy: TOrder[] = Array.from({ length: 10 }, () => {
      return {
        price: faker.commerce.price(undefined, 499),
        amount: faker.finance.amount(undefined, 100),
      };
    });

    const sell: TOrder[] = Array.from({ length: 10 }, () => {
      return {
        price: faker.commerce.price(500),
        amount: faker.finance.amount(undefined, 100),
      };
    });

    return {
      buy: buy.concat(this.userOrder.buy),
      sell: sell.concat(this.userOrder.sell),
    };
  }

  subscribe(callback: TOrderBookFunction) {
    this.subscribers.push(callback);
    callback(this.currentData);
    return () => {
      const index = this.subscribers.findIndex(
        (subscriber: TOrderBookFunction) => {
          return subscriber === callback;
        }
      );
      this.unSubscribe(index);
    };
  }

  unSubscribe(unSubscribeIndex: number) {
    this.subscribers = this.subscribers.filter(
      (_, index) => index !== unSubscribeIndex
    );
  }

  destroy() {
    window.clearInterval(this.intervalId);
    this.subscribers = [];
  }
}

export const orderBookStream = new OrderBookStream();
