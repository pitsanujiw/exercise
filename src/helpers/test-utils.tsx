import { render, RenderResult } from "@testing-library/react";
import React, { ReactElement } from "react";

const customRender = (ui: ReactElement, options = {}): RenderResult =>
  render(ui, { ...options });

export * from "@testing-library/react";

export { customRender };
