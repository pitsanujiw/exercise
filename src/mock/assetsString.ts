const assetsString = [
  {
    priceUsd: '33564.4851749769796344',
    timestamp: 1611986400000,
    date: 'Sat, 30 Jan 2021 06:00:00 GMT'
  },
  {
    priceUsd: '33459.0897814552902371',
    timestamp: 1611988200000,
    date: 'Sat, 30 Jan 2021 06:30:00 GMT'
  },
  {
    priceUsd: '33782.0757741344458237',
    timestamp: 1611990000000,
    date: 'Sat, 30 Jan 2021 07:00:00 GMT'
  }
]

export { assetsString }
