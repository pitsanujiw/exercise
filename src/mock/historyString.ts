const historyString = {
  priceUsd: '33782.0757741344458237',
  timestamp: 1611990000000,
  date: 'Sat, 30 Jan 2021 07:00:00 GMT'
}

export { historyString }
