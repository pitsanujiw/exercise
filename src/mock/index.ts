export * from './historyString';
export * from './assetsString';
