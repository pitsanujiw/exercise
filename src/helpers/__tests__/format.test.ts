import { Format, numberFormatter } from "helpers";
import { assetsString, historyString } from "mock";

describe("Format", () => {
  it("Should return correct big number format", () => {
    expect(Format.bigNumber(12345678910)).toMatchSnapshot();
  });

  it("Should return correct currency format", () => {
    expect(Format.currency(6789)).toMatchSnapshot();
  });

  it("Should return correct percent format", () => {
    expect(Format.percent(25)).toMatchSnapshot();
  });

  it("Should return correct number data in an array", () => {
    expect(Format.toNumbers(assetsString, ["priceUsd"])).toMatchSnapshot();
  });

  it("Should return correct to number format", () => {
    expect(Format.toNumber(historyString, "priceUsd")).toMatchSnapshot();
  });
});

describe("numberFormatter", () => {
  it("Should return correct format", () => {
    expect(numberFormatter.format(12345678910)).toMatchSnapshot();
  });
});
