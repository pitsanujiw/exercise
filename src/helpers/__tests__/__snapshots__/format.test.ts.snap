// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Format Should return correct big number format 1`] = `"$12.35b"`;

exports[`Format Should return correct currency format 1`] = `"$6,789.00"`;

exports[`Format Should return correct number data in an array 1`] = `
Array [
  Object {
    "date": "Sat, 30 Jan 2021 06:00:00 GMT",
    "priceUsd": 33564.48517497698,
    "timestamp": 1611986400000,
  },
  Object {
    "date": "Sat, 30 Jan 2021 06:30:00 GMT",
    "priceUsd": 33459.08978145529,
    "timestamp": 1611988200000,
  },
  Object {
    "date": "Sat, 30 Jan 2021 07:00:00 GMT",
    "priceUsd": 33782.07577413444,
    "timestamp": 1611990000000,
  },
]
`;

exports[`Format Should return correct percent format 1`] = `"25.00%"`;

exports[`Format Should return correct to number format 1`] = `33782.07577413444`;

exports[`numberFormatter Should return correct format 1`] = `"12,345,678,910"`;
